vcpkg overview page and article topic metadata fixes (#183)
microsoft/vcpkg-docs: vcpkg/get_started/overview.md

## Why vcpkg?

- Over 2200 open-source libraries to choose from, or bring your own
- Consistent, cross-platform experience for Windows, macOS, and Linux
- Dependencies can be built from source as needed with over 70 default build configurations (or customize your own builds)
- Avoid ABI incompatibilities between packages and your consuming project
- Prevent version conflicts and diamond problems in your dependency graph
- Enterprise-ready tool relied by C/C++ developers across many industries, including: finance, gaming, embedded/IoT, semiconductors, defense, and at Microsoft
- For MSBuild and CMake users: automatic integration with your build environment for a seamless dependency acquisition experience

## vcpkg features

### Ports and triplets

> A [port](../concepts/ports.md) in vcpkg is a versioned build recipe that produces a package. The most common type of package is a C/C++ library consisting of headers, source code, and binaries.

Reviewer: ~build~

> A [triplet](../users/triplets.md) captures the target build environment (cpu, os, compiler, runtime, etc.) in a single, convenient name. vcpkg provides over 70 triplets by default, but you can also define your own.

Reviewer: > (cpu, os, compiler, runtime, etc.)

I suggest adding `library type` to this list.

To install a package on your system, vcpkg runs the port's recipe file (a CMake script). The script may contain steps to download source code or run a build in your system. During builds, vcpkg uses the information in your triplet to make sure that the produced package matches your desired configuration.

### Binary caching

> While vcpkg builds libraries from source whenever it's necessary, you can back up your built packages in a [binary cache](../consume/binary-caching-overview.md). This allows other developer machines or continuous integration runs to reference these prebuilt packages without running a new build every time. vcpkg intelligently determines if a rebuild is necessary by checking if the cache already contains a valid existing package when it attempts to install a package.

Reviewer: "back up" has the wrong connotation -- it sounds like "critical persistant storage for disaster recovery".

> While vcpkg builds libraries from source whenever it's necessary, you can back up your built packages in a [binary cache](../consume/binary-caching-overview.md). This allows other developer machines or continuous integration runs to reference these prebuilt packages without running a new build every time. vcpkg intelligently determines if a rebuild is necessary by checking if the cache already contains a valid existing package when it attempts to install a package.

Reviewer: > This allows other developer machines...

We shouldn't use "other" here, because it is implying the scenario of sharing a read-write binary cache between multiple developers (a poor security practice). Instead, we should say "This allows developer machines or other continuous integration runs to reference these prebuilt packages...."

### Manifests

You can declare your direct dependencies and add optional features or version constraints in a [manifest file](../users/manifests.md).

### Versioning

> vcpkg has a unique way of handling [package versions](../users/versioning.concepts.md). Your manifest file can reference a single, baseline version set by default. The baseline references a combination of ports, each at a specific version, that have been tested for compatibility against each other. Furthermore, you can set a minimum version for a package or override a package version to get an exact match.

Reviewer: I think this paragraph could be reworded a bit to better emphasize "why" and "what" over "how". The purpose of baselines is to give you hassle-free, conflict-free dependency management with full reproducibility, while still giving you advanced control by pinning individual versions.

### Registries

> A [registry](../users/registries.md) is a catalog of ports and available versions that a vcpkg user can install. vcpkg provides a public registry of open-source libraries by default, but you can also define your own registries for custom libraries.

Reviewer: > ... default, but you can also define your own registries for custom libraries.

Suggestion: "... default and you can create your own registry for customizations, patches, or private libraries."

### Asset caching

> [Asset caching](../users/assetcaching.md) allows vcpkg to use download-mirrors to upload and restore assets such as source code and build tools. This is useful for continuity of business, especially for open-source dependencies that have assets on third party servers. This also provides the developer protection in case upstream assets are tampered with unexpectedly.

Reviewer: Lead with the point:

"Asset caching allows vcpkg to work in air-gapped and offline environments, ensuring continuity of business even if a remote host goes down or is compromised."

Then, we can tease how that's accomplished.

NuGet is a .NET package manager that has often been used for C/C++ development, particularly for MSBuild solutions that also contain .NET projects. As a general rule, the Microsoft C++ team does not recommend using NuGet for C/C++ development, as NuGet has several limitations:

- **Compilation flavors**. As NuGet cannot build packages from source on the fly, it is necessary to provide prebuilt binaries to match all possible Application Binary Interface (ABI) restrictions for all users. The user is responsible for building packages themselves and ensuring this is done correctly. It is also difficult to distinguish between binaries due to the lack of relevant metadata, resulting in the user having to do quirky things like putting architecture, operating system, and compiler information in the package name, without being able to enforce such constraints during package acquisition.
- **Binary vs. source**. Very closely tied to the first point, NuGet is designed from the ground up to provide relatively small, prebuilt binaries. Due to the nature of native code, developers need to have access to the source code to ensure ABI compatibility, performance, integrity, and debuggability.

> - **Versioning limitations**. While NuGet packages are versioned, it is up to the user to decide on appropriate versions of packages to take dependencies on. The pressure is on the user to solve version conflicts between dependencies and the consuming project.

Reviewer: This point looks like a positive from a certain perspective: "Oh, I'm a smart, advanced user, I like having lots of control! Does vcpkg not let me decided on the appropriate versions of packages to use?" Our versioning system is slightly easier to use, but I wouldn't describe NuGet as having "Versioning limitations".

- **No NuGet PackageReference support**. NuGet PackageReference is not supported for .vcxproj files and there is no plan to add it in the future due to technical and architectural differences between C++ and .NET MSBuild projects. This means NuGet C++ users won't benefit from features like the use of the global cache and referencing dependencies in simple MSBuild terms with access to conditional logic.

> - **Environment changes**. As NuGet C++ packages must be manually compiled and packaged, any time the host or target environment changes this work must be repeated. On the other hand, vcpkg can build packages from source for a variety of operating systems and build systems.

Reviewer: This sounds an awful lot like "Compilation flavors". I'd replace this section with talking about transitive ABI tracking and how vcpkg can insure that when boost updates, every library using boost also rebuilds.

ms.topic: overview
author: AugP
ms.author: aupopa
ms.date: 10/22/2023
---

# vcpkg overview

> vcpkg (always spelled lowercase) is a free and open-source C/C++ package manager maintained by Microsoft and the C++ community. Initially launched in 2016 as a tool for assisting developers in migrating their projects to newer versions of Visual Studio, vcpkg has evolved into a cross-platform tool used by developers on Windows, macOS, and Linux with a large catalog of open-source libraries and enterprise-ready features. vcpkg is designed to not get in the way of your development process with support for a variety of build and project systems. As a C++ tool at heart, vcpkg is written primarily in C++ and CMake. It is designed to address common pain points for C/C++ developers in ways generally ignored by other package managers.

Reviewer 2: ```suggestion
vcpkg (always spelled lowercase) is a free and open-source C/C++ package manager maintained by Microsoft and the C++ community. Initially launched in 2016 as a tool for assisting developers in migrating their projects to newer versions of Visual Studio, vcpkg has evolved into a cross-platform tool used by developers on Windows, macOS, and Linux. vcpkg has a large catalog of open-source libraries and enterprise-ready features designed to facilitate your development process with support for a variety of build and project systems. As a C++ tool at heart, vcpkg is written primarily in C++ and CMake. It is designed to address common pain points for C/C++ developers in ways generally ignored by other package managers.
```

Split the second sentence since it runs too long. Also changed negative language ("not get in your way") to something more positive. 

Also, that last sentence seems unnecessary, since we don't mention what are the "pain points" that other package managers ignore that vcpkg does not. I think the list in the next paragraph makes a better argument to sell you the idea of vcpkg.

## Why vcpkg?

- Over 2200 open-source libraries to choose from, or bring your own
- Consistent, cross-platform experience for Windows, macOS, and Linux
- Dependencies can be built from source as needed with over 70 default build configurations (or customize your own builds)
- Avoid ABI incompatibilities between packages and your consuming project
- Prevent version conflicts and diamond problems in your dependency graph
- Enterprise-ready tool relied by C/C++ developers across many industries, including: finance, gaming, embedded/IoT, semiconductors, defense, and at Microsoft
- For MSBuild and CMake users: automatic integration with your build environment for a seamless dependency acquisition experience

## vcpkg features

### Ports and triplets

> A [port](../concepts/ports.md) in vcpkg is a versioned build recipe that produces a package consisting of a set of files. The most common package that can get produced when a port is installed is a C/C++ library consisting of headers, source code, and binaries.

Reviewer 2: ```suggestion
A [port](../concepts/ports.md) in vcpkg is a versioned build recipe that produces a package. The most common type of package is a C/C++ library consisting of headers, source code, and binaries.
```

A [triplet](../users/triplets.md) captures the target build environment (cpu, os, compiler, runtime, etc.) in a single, convenient name. vcpkg provides over 70 triplets by default, but you can also define your own.

> To install a package on your system, vcpkg runs the script specified in the port's *portfile.cmake* file. This can include downloading source code and running a build. To run the build, it will use the triplet information to make sure that the final package matches the desired configuration.

Reviewer 2: ```suggestion
To install a package on your system, vcpkg runs the port's recipe file (a CMake script). The script may contain steps to download source code or run a build in your system. During builds, vcpkg uses the information in your triplet to make sure that the produced package matches your desired configuration.
```

Since this is an overview, I don't think we need to go into detail about `portfile.cmake`, just calling it a script should be enough.
